### PR TITLE
Update boot_utils for variable-sized mcuboot image trailer

### PR DESCRIPTION
**Pull request by mbolivar**
Upstream mcuboot has changed the image trailer format. This series attempts to keep zephyr-fota-hawkbit up to date.

Hoping for testing from @Hashcode on K64F.

### CODE REVIEW

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

**Comment by gatek33p3r** [comment]
Bootloader Build Failed.

**Comment by EmbeddedAndroid** [comment]
rebuild

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

File: src/boot_utils.c [hunk 1]
```diff
@@ -17,21 +17,71 @@
...
+ */
+
+#define TRAILER_SWAP_STATUS_SIZE	(128 * FLASH_MIN_WRITE_SIZE * 3)
+#define TRAILER_COPY_DONE_SIZE		FLASH_MIN_WRITE_SIZE
+#define TRAILER_IMAGE_OK_SIZE		FLASH_MIN_WRITE_SIZE
+
+#define TRAILER_COPY_DONE		0x01
+#define TRAILER_PADDING			0xff
+
+static const uint32_t boot_img_magic[4] = {
```
Comment (rsalveti) [inline src/boot_utils.c]: Mind adding an extra comment pointing to mcuboot for the original values?

File: src/boot_utils.c [hunk 2]
```diff
@@ -41,37 +91,43 @@ uint8_t boot_status_read(void)
...
-	img_trailer.bit_copy_start = BOOT_IMG_MAGIC;
-	img_trailer.bit_copy_done = 0xff;
-	img_trailer.bit_img_ok = 0xff;
-
-	offset = FLASH_BANK1_OFFSET + FLASH_BANK_SIZE -
-				sizeof(struct boot_img_trailer);
+	magic_offset = boot_trailer_magic(FLASH_BANK1_OFFSET);
+	copy_done_offset = boot_trailer_copy_done(FLASH_BANK1_OFFSET);
+	memset(copy_done, TRAILER_PADDING, sizeof(copy_done));
+	copy_done[0] = TRAILER_COPY_DONE;
```
Comment (rsalveti) [inline src/boot_utils.c]: Why do we need to set copy done here?

File: src/boot_utils.c (hunk 1 above)
Comment (mbolivar) [inline src/boot_utils.c]: There is a comment above noting that the definitions here are from mcuboot. I'll rename this to "trailer_img_magic" so it's more clear that it's part of that section.

File: src/boot_utils.c (hunk 2 above)
Comment (mbolivar) [inline src/boot_utils.c]: I'm just reproducing the existing behavior (set the magic number and copy done fields) with the new layout.

Did the way mcuboot treats these fields change? Or maybe we're planning on setting up the fields properly in zep2newt so the trailer is flashed by the OTA process?

Comment (rsalveti) [inline src/boot_utils.c]: The existing behavior is setting copy_done to 0xff, not TRAILER_COPY_DONE (0x01), unless I'm missing something here.

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

File: src/boot_utils.c (hunk 2 above)
Comment (mbolivar) [inline src/boot_utils.c]: Hmm, right. I was going off the log message and didn't parse the magic numbers correctly.

This is actually setting the magic number and clearing both copy_done and image_ok. I need to fix this here and update the next commit to match.

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

**Comment by mike-scott** [comment]
Tested successfully on K64F.  I haven't take a good look at the patches yet.

File: src/boot_utils.c [hunk 3]
```diff
@@ -41,37 +88,49 @@ uint8_t boot_status_read(void)
...
+	memset(image_ok, 0xff, sizeof(image_ok));
 
-	img_trailer.bit_copy_start = BOOT_IMG_MAGIC;
-	img_trailer.bit_copy_done = 0xff;
-	img_trailer.bit_img_ok = 0xff;
+	OTA_INFO("Setting trailer magic number; clearing image_ok and copy_done\n");
 
-	offset = FLASH_BANK1_OFFSET + FLASH_BANK_SIZE -
-				sizeof(struct boot_img_trailer);
+	flash_write_protection_set(flash_dev, false);
```
Comment (mike-scott) [inline src/boot_utils.c]: Do we need to call flash_write_protection_set false/true around each of the flash_write calls below or should we call it once at the beginning and once at the end?

File: src/boot_utils.c [hunk 4]
```diff
@@ -41,37 +88,49 @@ uint8_t boot_status_read(void)
...
+	flash_write(flash_dev, magic_offset, trailer_img_magic,
```
Comment (mike-scott) [inline src/boot_utils.c]: After discussing on Slack we decided we could remove all of the image_magic related code.  We can use our image creation tools (zep2newt) to add that for us with the --bit option.  Less code here to maintain the better, IMHO.

File: src/boot_utils.c (hunk 3 above)
Comment (mbolivar) [inline src/boot_utils.c]: Unfortunately, we really do have to do it this way.

```
/**
 *  @brief  Enable or disable write protection for a flash memory
 *
 *  This API is required to be called before the invocation of write or erase
 *  API. Please note that on some flash components, the write protection is
 *  automatically turned on again by the device after the completion of each
 *  write or erase calls. Therefore, on those flash parts, write protection needs
 *  to be disabled before each invocation of the write or erase API. Please refer
 *  to the sub-driver API or the data sheet of the flash component to get details
 *  on the write protection behavior.
 *
 *  @param  dev             : flash device
 *  @param  enable          : enable or disable flash write protection
 *
 *  @return  0 on success, negative errno code on fail.
 */
static inline int flash_write_protection_set(struct device *dev, bool enable)
```

File: src/boot_utils.c (hunk 4 above)
Comment (mbolivar) [inline src/boot_utils.c]: Agreed.

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

File: src/boot_utils.c [hunk 5]
```diff
@@ -17,21 +17,68 @@
...
 
+/*
+ * Helpers for image trailer, as defined by mcuboot.
+ */
+
+#define TRAILER_SWAP_STATUS_SIZE	(128 * FLASH_MIN_WRITE_SIZE * 3)
+#define TRAILER_COPY_DONE_SIZE		FLASH_MIN_WRITE_SIZE
+#define TRAILER_IMAGE_OK_SIZE		FLASH_MIN_WRITE_SIZE
+
+static const uint32_t trailer_img_magic[4] = {
```
Comment (mike-scott) [inline src/boot_utils.c]: Let's remove the actual values from source, but keep the definition of what the tailer_img_magic is (uint32_t[4])

**Comment by mike-scott** [comment]
Confirmed this latest patch works on FRDM-K64F

File: src/boot_utils.c (hunk 5 above)
Comment (mbolivar) [inline src/boot_utils.c]: Do you mean `typedef struct trailer_img_magic { uint32_t magic[4]; } trailer_img_magic_t;`? If not, I'm not sure what you're asking for here.

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

File: src/boot_utils.c (hunk 5 above)
Comment (mbolivar) [inline src/boot_utils.c]: Done

**Comment by gatek33p3r** [comment]
Checkpatch Suceeded.

**Comment by gatek33p3r** [comment]
Application Build Succeeded.

**Comment by gatek33p3r** [comment]
Bootloader Build Succeeded.

**Pull request merged**